sktime/forecasting/compose/_pipeline.py

```diff
@@ -1583,10 +1583,11 @@ def _get_X_for_fcX(self, X):
             return None
 
         if ixx == "complement":
-            X_for_fcX = X.drop(columns=self.columns)
-            if X_for_fcX.shape[1] < 1:
-                return None
-            return X_for_fcX
+            all_columns = X.columns
```

Comment:
can you explain why this fixes the problem? I don't see a difference in behaviour, except in handling the case where all columns are dropped.

Should we not still return `None` in that case, or was that the problem?

Reply:
This was the problem as the columns were being dropped for the case where `complement` was passed and the `self.columns` is not `None`  even if some columns were present in `self` but not in `X` the `all_columns` is used to compute the complemented columns as used on the line below, we can also add an `errors='ignore'` to the `X.drop(columns=self.columns, errors='ignore')` to avoid the current issue. Computing the complemented columns first makes sense but will sometimes return an empty dataFrame and I am unsure if it is the correct practice for dataFrames with many columns as well. Should we keep the current implementation and add `errors='ignore'` in the `drop` method?

Reply:
And I agree with returning `None` when all columns are dropped.

Reply:
oh, I see - can I check whether I understand what you are saying:

* the failure case is when the user passes `columns` that have some columns which do not appear in `X`
* this causes a failure in `drop`

> I am unsure if it is the correct practice for dataFrames with many columns 

Can you expand on this? Why should the number of columns matter?

> Should we keep the current implementation and add `errors='ignore'` in the `drop` method?

Yes, I think that might be the most compact solution, now that you've explained it. Assuming I got your points right.

Reply:
> oh, I see - can I check whether I understand what you are saying:
> 
> * the failure case is when the user passes `columns` that have some columns which do not appear in `X`
> * this causes a failure in `drop`
> 
> > I am unsure if it is the correct practice for dataFrames with many columns
> 
> Can you expand on this? Why should the number of columns matter?

The reason why the number of `columns` should matter is that when computing the complemented columns we are introducing overhead due to list comprehension and usage of `loc`. Let's say the length of `X.columns` as `m` and `n` as length of `self.columns`, the complexity for using `drop` should be of order `O(m+n)` where as using list comprehension to compute `complemented_columns` should be `O(m.n)` and `loc` function is of order `O(n)`

> > Should we keep the current implementation and add `errors='ignore'` in the `drop` method?
> 
>> Yes, I think that might be the most compact solution, now that you've explained it. Assuming I got your points right.

Then I am going to keep the drop function and test if it works.

Reply:
Makes sense - so `drop` has the best complexity, right? Let's keep it then.

Reply:
Sure